paasta_tools/apply_external_resources.py

```diff
@@ -0,0 +1,76 @@
+#!/opt/venvs/paasta-tools/bin/python
+import os
+import sys
+from filecmp import cmp
+from shutil import copy
+from subprocess import CalledProcessError
+from subprocess import run
+from traceback import print_exc
+
+APPLIED_DIRECTORY = ".applied"
+
+
+# This script expects the KUBECONFIG environment variable to be set correctly
+def main(puppet_resource_root: str) -> int:
+    exit_code = 0
+    applied_resource_root = os.path.join(puppet_resource_root, APPLIED_DIRECTORY)
+
+    # Loop through everything in the puppet resource path
+    for root, dirs, files in os.walk(puppet_resource_root):
+        # modifying the 'dirs' variable in-place will update the order that os.walk visits things
+        if APPLIED_DIRECTORY in dirs:
+            dirs.remove(APPLIED_DIRECTORY)
+        dirs.sort()  # Need to apply things in the correct order
+
+        # Check to see if there's a difference between what Puppet created and
+        # what's been previously applied
+        for filename in sorted([f for f in files if f.endswith(".yaml")]):
+            path = os.path.join(root, filename)
+            applied_path = os.path.join(
+                applied_resource_root, path[len(puppet_resource_root) + 1 :]
+            )
+            print(f"comparing {path} and {applied_path}")
+            if not os.path.exists(applied_path) or not cmp(
+                path, applied_path, shallow=False
+            ):
+                # This is idempotent; if something gets out of sync and a resource gets applied
+                # a second time, kubectl just won't make any changes
+                try:
+                    run(["kubectl", "apply", "-f", path], check=True)
+                    os.makedirs(applied_path[: -len(filename)], exist_ok=True)
+                    copy(path, applied_path)
+                except CalledProcessError:
+                    print(f"There was a problem applying {path}:\n")
+                    print_exc(
+                        file=sys.stdout
+                    )  # keep all messages on the same stream so they're in order
+                    exit_code = 1
+                    continue
+
+    # Loop through all the files that have been previously applied and see
+    # if Puppet has removed any of them
+    for root, dirs, files in os.walk(applied_resource_root):
+        dirs.sort(reverse=True)  # for deleting things, we need to go in reverse order
+        for filename in sorted([f for f in files if f.endswith(".yaml")], reverse=True):
```

Comment:
The same as above: `dirs` doesn't seem to be used, full paths for files from all directories should be sorted.

Reply:
see above.